src/goto-programs/remove_exceptions.cpp

```diff
@@ -0,0 +1,583 @@
+
+  if(goto_program.empty())
+    return;
+
+  Forall_goto_program_instructions(instr_it, goto_program)
+  {
+    // is this a handler
+    if(instr_it->type==CATCH && instr_it->code.find(ID_handler)!=get_nil_irep())
+    {
+      // retrieve the exception variable
+      const irept &exception=instr_it->code.find(ID_handler);
+
+      if(symbol_table.has_symbol(id2string(function_id)+EXC_SUFFIX))
+      {
+        symbolt &function_symbol=symbol_table.lookup(id2string(function_id)+
+                                                    EXC_SUFFIX);
+        // next we reset the exceptional return to NULL
+        symbol_exprt lhs_expr_null;
+        lhs_expr_null.set_identifier(id2string(function_id)+EXC_SUFFIX);
+        lhs_expr_null.type()=function_symbol.type;
+        exprt rhs_expr_null;
+        rhs_expr_null=null_pointer_exprt(pointer_typet());
+
+        // add the assignment
+        goto_programt::targett t_null=goto_program.insert_after(instr_it);
+        t_null->make_assignment();
+        t_null->source_location=instr_it->source_location;
+        t_null->code=code_assignt(
+          lhs_expr_null,
+          typecast_exprt(rhs_expr_null, lhs_expr_null.type()));
+        t_null->function=instr_it->function;
+
+        // add the assignment exc=f#exception_value
+        symbol_exprt rhs_expr_exc;
+        rhs_expr_exc.set_identifier(id2string(function_id)+EXC_SUFFIX);
+        rhs_expr_exc.type()=function_symbol.type;
+
+        const exprt &lhs_expr_exc=static_cast<const exprt &>(exception);
+        goto_programt::targett t_exc=goto_program.insert_after(instr_it);
+        t_exc->make_assignment();
+        t_exc->source_location=instr_it->source_location;
+        t_exc->code=code_assignt(
+          typecast_exprt(lhs_expr_exc, rhs_expr_exc.type()),
+          rhs_expr_exc);
+        t_exc->function=instr_it->function;
+      }
+
+      instr_it->make_skip();
+    }
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_exceptionst::add_gotos_throw
+
+Inputs:
+
+Outputs:
+
+Purpose: instruments each throw with conditional GOTOS to the 
+         corresponding exception handlers
+
+\*******************************************************************/
+void remove_exceptionst::add_throw_gotos(
+  goto_functionst::function_mapt::iterator func_it,
+  goto_programt::instructionst::iterator instr_it,
+  stack_catcht stack_catch)
+{
+  assert(instr_it->type==THROW);
+
+  goto_programt &goto_program=func_it->second.body;
+
+  assert(instr_it->code.operands().size()==1);
+
+  // make sure that these are always forward gotos
+  assert(!instr_it->is_backwards_goto());
+
+  // jump to the end of the function
+  // this will appear after the GOTO-based dynamic dispatch below
+  goto_programt::targett t_end=goto_program.insert_after(instr_it);
+  goto_programt::targett new_state;
+  // find the end of the function
+  for(goto_programt::instructionst::iterator inner_it=instr_it;
```

Comment:
Why do you have to iterate to the end? 
Isn't it sufficient to just use new_state as iterator?